src/browser/fake_network.js

```diff
@@ -43,6 +58,220 @@ function iptolong(parts) {
     return parts[0] << 24 | parts[1] << 16 | parts[2] << 8 | parts[3];
 }
 
+class Uint8Stream
+{
+    /**
+     * @param {number} initial_capacity
+     * @param {number} maximum_capacity
+     */
+    constructor(initial_capacity, maximum_capacity)
+    {
+        this.maximum_capacity = maximum_capacity;
+        this.tail = 0;
+        this.head = 0;
+        this.length = 0;
+        this.buffer = new Uint8Array(initial_capacity);
+    }
+
+    /**
+     * @param {number} capacity
+     */
+    resize(capacity)
+    {
+        const new_buffer = new Uint8Array(capacity);
+        this.peek(new_buffer, this.length);
+        this.tail = 0;
+        this.head = this.length;
+        this.buffer = new_buffer;
+    }
+
+    /**
+     * @param {Uint8Array} src_array
+     */
+    write(src_array)
+    {
+        const src_length = src_array.length;
+        const total_length = this.length + src_length;
+        let capacity = this.buffer.length;
+        if(capacity < total_length) {
+            while(capacity < total_length) {
+                capacity *= 2;
+            }
+            if(this.maximum_capacity && capacity > this.maximum_capacity) {
+                console.error("stream capacity overflow in Uint8Stream.write()");
+                return;
+            }
+            this.resize(capacity);
+        }
+        const buffer = this.buffer;
+
+        /*
+        let head = this.head;
+        for(let i=0; i<src_length; ++i) {
+            buffer[head] = src_array[i];
+            head = (head + 1) % capacity;
+        }
+        this.head = head;
+        this.length += src_length;
+        */
+        const new_head = this.head + src_length;
+        if(new_head > capacity) {
+            const i_split = capacity - this.head;
+            buffer.set(src_array.subarray(0, i_split), this.head);
+            buffer.set(src_array.subarray(i_split));
+        }
+        else {
+            buffer.set(src_array, this.head);
+        }
+        this.head = new_head % capacity;
+        this.length += src_length;
+    }
+
+    /**
+     * @param {Uint8Array} dst_array
+     * @param {number} length
+     */
+    peek(dst_array, length)
+    {
```

Comment:
I think we could write `let length = dst_array.length` here and remove the `length` parameter.

Reply:
Yes, I had overlooked that and changed it.